File: ip-validator/typescript/src/ipValidator.ts
export function isValid(input: string): boolean {
  if (input.length === 0) return false;

  const octets = input.split('.');
  if (octets.length !== 4) return false;

  for (let i = 0; i < 4; i++) {
    const value = parseOctet(octets[i]!);
    if (value === null) return false;
    if (i === 3 && (value === 0 || value === 255)) return false;
  }

  return true;
}

function parseOctet(octet: string): number | null {
  if (octet.length === 0) return null;
  if (octet.length > 1 && octet[0] === '0') return null;

  let value = 0;
  for (const c of octet) {
    if (c < '0' || c > '9') return null;
    value = value * 10 + (c.charCodeAt(0) - '0'.charCodeAt(0));
    if (value > 255) return null;
  }
  return value;
}

File: ip-validator/python/src/ip_validator/ip_validator.py
def is_valid(input: str) -> bool:
    if not input:
        return False

    octets = input.split(".")
    if len(octets) != 4:
        return False

    for i, octet in enumerate(octets):
        value = _parse_octet(octet)
        if value is None:
            return False
        if i == 3 and value in (0, 255):
            return False

    return True


def _parse_octet(octet: str) -> int | None:
    if len(octet) == 0:
        return None
    if len(octet) > 1 and octet[0] == "0":
        return None

    value = 0
    for c in octet:
        if not ("0" <= c <= "9"):
            return None
        value = value * 10 + (ord(c) - ord("0"))
        if value > 255:
            return None
    return value
